### PR TITLE
respect user choice of lib/bin over heuristics

### LINKED ISSUE #9333: cargo init --bin creates a lib crate when crate source file is present
**Problem**

`cargo init` has a nice feature, in that if the file `foo.rs` exists in the directory `foo` that is being initialized, no `src` directory will be created and instead `name = "foo"` and `path = "foo.rs"` will be placed in `Cargo.toml`.

Unfortunately, `cargo init --bin` still produces a `lib` crate in this case.

**Steps**

1. `mkdir foo`
2. `cd foo`
3. `touch foo.rs`
4. `cargo init --bin`

The resulting `Cargo.toml` is

```
[package]
name = "foo"
version = "0.1.0"
authors = ["Me <me@example.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]

[lib]
name = "foo"
path = "foo.rs"
```

I would strongly prefer getting `[[bin]]` instead of `[lib]` here.

**Notes**

Output of `cargo version`: `cargo 1.51.0 (43b129a20 2021-03-16)`

### PR DESCRIPTION
**Pull request by Aelnor**
This one fixes #9333

### CODE REVIEW

**Comment by Aelnor** [comment]
Sorry, the PR is premature,  I've found a bug here

File: tests/testsuite/init.rs [hunk 1]
```diff
@@ -598,3 +598,68 @@ mod tests {
...
+
+#[cargo_test]
+fn creates_binary_when_instructed_and_has_lib_file_no_warning() {
+    let path = paths::root().join("foo");
+    fs::create_dir(&path).unwrap();
+    fs::write(path.join("foo.rs"), "fn not_main() {}").unwrap();
+    cargo_process("init --bin")
+        .cwd(&path)
+        .with_stderr(
+            "[WARNING] file 'foo.rs' seems to be a library file\n[CREATED] binary (application) package"
```
Comment (Aelnor) [inline tests/testsuite/init.rs:610]: I'm not sure in this case we should issue a warning. It means there's no `fn main` most likely

File: tests/testsuite/init.rs [hunk 2]
```diff
@@ -598,3 +598,68 @@ mod tests {
...
+
+#[cargo_test]
+fn creates_library_when_instructed_and_has_bin_file() {
+    let path = paths::root().join("foo");
+    fs::create_dir(&path).unwrap();
+    fs::write(path.join("foo.rs"), "fn main() {}").unwrap();
+    cargo_process("init --lib")
+        .cwd(&path)
+        .with_stderr(
+            "[WARNING] file 'foo.rs' seems to be a binary (application) file\n[CREATED] library package"
```
Comment (weihanglo) [inline tests/testsuite/init.rs:627]: Prefer this multi-line style for being consistent with other tests

```suggestion
            "\
[WARNING] file 'foo.rs' seems to be a binary (application) file
[CREATED] library package
"
```

File: src/cargo/ops/cargo_new.rs [hunk 3]
```diff
@@ -62,19 +62,25 @@ pub struct NewOptions {
...
+        self == NewProjectKind::Auto
+    }
 }
 
 impl fmt::Display for NewProjectKind {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match *self {
             NewProjectKind::Bin => "binary (application)",
             NewProjectKind::Lib => "library",
+            NewProjectKind::Auto => "auto-select type",
```
Comment (weihanglo) [inline src/cargo/ops/cargo_new.rs:83]: `NewProjectKind::Auto` seems quite unnecessary at this time. Perhaps we can add a `auto_detect` field for `NewOptions` struct instead.

File: src/cargo/ops/cargo_new.rs [hunk 4]
```diff
@@ -399,20 +424,22 @@ pub fn new(opts: &NewOptions, config: &Config) -> CargoResult<()> {
         )
     }
 
+    let kind = match opts.kind {
+        NewProjectKind::Bin => NewProjectKind::Bin,
+        NewProjectKind::Auto => NewProjectKind::Bin,
+        _ => NewProjectKind::Lib,
+    };
```
Comment (weihanglo) [inline src/cargo/ops/cargo_new.rs:431]: These can be deleted If removing `NewProjectKind::Auto`.

File: src/cargo/ops/cargo_new.rs [hunk 5]
```diff
@@ -445,14 +472,34 @@ pub fn init(opts: &NewOptions, config: &Config) -> CargoResult<()> {
...
+        src_paths_types.push(plan_new_source_file(has_bin, name.to_string()));
+    } else if src_paths_types.len() == 1 && !src_paths_types.iter().any(|x| x.bin == has_bin) {
+        // we've found the only file and it's not the type user wants. Change the type and warn
+        let file_type = if src_paths_types[0].bin {
+            NewProjectKind::Bin.to_string()
+        } else {
+            NewProjectKind::Lib.to_string()
+        };
+        config.shell().warn(format!(
+            "file '{}' seems to be a {} file",
```
Comment (weihanglo) [inline src/cargo/ops/cargo_new.rs:488]: IIRC there is a convention to use `` in error message.

```suggestion
            "file `{}` seems to be a {} file",
```

File: src/cargo/ops/cargo_new.rs [hunk 6]
```diff
@@ -445,14 +472,34 @@ pub fn init(opts: &NewOptions, config: &Config) -> CargoResult<()> {
...
-        // --bin option may be ignored if lib.rs or src/lib.rs present
-        // Maybe when doing `cargo init --bin` inside a library package stub,
-        // user may mean "initialize for library, but also add binary target"
+        src_paths_types.push(plan_new_source_file(has_bin, name.to_string()));
+    } else if src_paths_types.len() == 1 && !src_paths_types.iter().any(|x| x.bin == has_bin) {
+        // we've found the only file and it's not the type user wants. Change the type and warn
+        let file_type = if src_paths_types[0].bin {
+            NewProjectKind::Bin.to_string()
+        } else {
+            NewProjectKind::Lib.to_string()
```
Comment (weihanglo) [inline src/cargo/ops/cargo_new.rs:485]: `NewProjectKind` has already implemented `Display` so need not to call `to_string` explicitly.

```suggestion
            NewProjectKind::Bin
        } else {
            NewProjectKind::Lib
```

**Review by weihanglo — COMMENTED** [review]
Just a folk from nowhere and saw your PR. Great job!

File: src/cargo/ops/cargo_new.rs (hunk 3 above)
Comment (Aelnor) [reply src/cargo/ops/cargo_new.rs]: Tbh, I like my variant a little bit more, but with your way the code becomes a little bit simpler, so let it be :)

File: src/cargo/ops/cargo_new.rs (hunk 4 above)
Comment (Aelnor) [reply src/cargo/ops/cargo_new.rs]: Done

File: src/cargo/ops/cargo_new.rs (hunk 5 above)
Comment (Aelnor) [reply src/cargo/ops/cargo_new.rs]: Done

File: tests/testsuite/init.rs (hunk 1 above)
Comment (ehuss) [reply tests/testsuite/init.rs]: The warning seems fine to me.  I imagine the user can figure things out afterwards.

**Comment by ehuss** [comment]
Looks great, thanks!

@bors r+

**Pull request merged**